scipy/stats/stats.py

```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    else:
+        samp_size = len(distx) - 1
+
+        # threshold to find connected region of significant local correlations
+        sig_connect = _threshold_mgc_map(stat_mgc_map, samp_size)
+
+        # maximum within the significant region
+        stat, opt_scale = _smooth_mgc_map(sig_connect, stat_mgc_map)
+
+    stat_dict = {"stat_mgc_map": stat_mgc_map,
+                 "opt_scale": opt_scale}
+
+    return stat, stat_dict
+
+
+def _threshold_mgc_map(stat_mgc_map, samp_size):
+    r"""
+    Finds a connected region of significance in the MGC-map by thresholding.
+
+    Parameters
+    ----------
+    stat_mgc_map : ndarray
+        All local correlations within `[-1,1]`.
+    samp_size : int
+        The sample size of original data.
+
+    Returns
+    -------
+    sig_connect : ndarray
+        A binary matrix with 1's indicating the significant region.
+    """
+    m, n = stat_mgc_map.shape
+
+    # parametric threshold, based on a threshold is estimated based on the
+    # normal distribution approximation
+    per_sig = 1 - (0.02 / samp_size)  # Percentile to consider as significant
+    threshold = samp_size * (samp_size - 3)/4 - 1/2  # Beta approximation
+    threshold = distributions.beta.ppf(per_sig, threshold, threshold) * 2 - 1
+
+    # take the max of threshold and local correlation at the maximal scale
+    threshold = max(threshold, stat_mgc_map[m - 1][n - 1])
+
+    # find the largest connected component of significant correlations
+    sig_connect = stat_mgc_map > threshold
+    if np.sum(sig_connect) > 0:
+        sig_connect, _ = measurements.label(sig_connect)
+        _, label_counts = np.unique(sig_connect, return_counts=True)
+
+        # skip the first element in label_counts, as it is count(zeros)
+        max_label = np.argmax(label_counts[1:]) + 1
+        sig_connect = sig_connect == max_label
+    else:
+        sig_connect = np.array([[False]])
+
+    return sig_connect
+
+
+def _smooth_mgc_map(sig_connect, stat_mgc_map):
+    """
+    Finds the smoothed maximal within the significant region R.
+
+    If area of R is too small it returns the last local correlation. Otherwise,
+    returns the maximum within significant_connected_region.
+
+    Parameters
+    ----------
+    sig_connect: ndarray
+        A binary matrix with 1's indicating the significant region.
+    stat_mgc_map: ndarray
+        All local correlations within `[-1,1]`.
+
+    Returns
+    -------
+    stat : float
+        The sample MGC statistic within `[-1, 1]`.
+    opt_scale: (float, float)
+        The estimated optimal scale as an `(x, y)` pair.
+    """
+
+    m, n = stat_mgc_map.shape
+
+    # default sample mgc and optimal scale to local corr at max scale
+    stat = stat_mgc_map[m - 1][n - 1]
+    opt_scale = [m, n]
+
+    if np.linalg.norm(sig_connect) != 0:
+        # proceed only when the connected region's area is sufficiently large
+        if np.sum(sig_connect) >= np.ceil(0.02 * max(m, n)) * min(m, n):
```

Comment:
another 0.02 value that deserves a comment/reference

Reply:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0